r/aws_batch_job_defintion: `_List_basic` acceptance test is broken

### Description

The test currently fails with the following on `main`:

```console
% make t K=batch T=TestAccBatchJobDefinition_List
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 main 🌿...
TF_ACC=1 go1.25.5 test ./internal/service/batch/... -v -count 1 -parallel 20 -run='TestAccBatchJobDefinition_List'  -timeout 360m -vet=off
2025/12/22 10:58:15 Creating Terraform AWS Provider (SDKv2-style)...
2025/12/22 10:58:15 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccBatchJobDefinition_List_basic
=== PAUSE TestAccBatchJobDefinition_List_basic
=== CONT  TestAccBatchJobDefinition_List_basic
    job_definition_list_test.go:29: Step 2/2 error running query checks: running terraform query command returned diagnostics: [{baseLogMessage:{Lvl:error Msg:Error: Error Listing Remote Resources Time:2025-12-22 10:58:26.493679 -0500 EST} Diagnostic:{Severity:error Summary:Error Listing Remote Resources Detail:An unexpected error occurred converting identity state. This is always an error in the provider. Please report the following to the provider developer:

        Error: state is nil, call SetId() on ResourceData first Range:0x140080199c0 Snippet:0x14003658b40}}]
--- FAIL: TestAccBatchJobDefinition_List_basic (13.65s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/batch      20.112s
```

### Important Facts and References

Discovered reviewing #45676

### Would you like to implement a relevant change?

No

--- a/.changelog/45694.txt
+++ b/.changelog/45694.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_batch_job_definition: Fix handling of logically deleted results in List
+```

--- a/internal/service/batch/job_definition.go
+++ b/internal/service/batch/job_definition.go
@@ -15,9 +15,6 @@ import (
 	"github.com/aws/aws-sdk-go-v2/service/batch"
 	awstypes "github.com/aws/aws-sdk-go-v2/service/batch/types"
 	"github.com/hashicorp/go-cty/cty"
-	"github.com/hashicorp/terraform-plugin-framework/list"
-	listschema "github.com/hashicorp/terraform-plugin-framework/list/schema"
-	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	sdkretry "github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
@@ -26,17 +23,13 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/conns"
 	"github.com/hashicorp/terraform-provider-aws/internal/enum"
 	"github.com/hashicorp/terraform-provider-aws/internal/errs"
-	"github.com/hashicorp/terraform-provider-aws/internal/errs/fwdiag"
 	"github.com/hashicorp/terraform-provider-aws/internal/errs/sdkdiag"
 	"github.com/hashicorp/terraform-provider-aws/internal/flex"
-	"github.com/hashicorp/terraform-provider-aws/internal/framework"
-	"github.com/hashicorp/terraform-provider-aws/internal/logging"
 	"github.com/hashicorp/terraform-provider-aws/internal/provider/sdkv2/importer"
 	"github.com/hashicorp/terraform-provider-aws/internal/retry"
 	"github.com/hashicorp/terraform-provider-aws/internal/sdkv2"
 	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
-	inttypes "github.com/hashicorp/terraform-provider-aws/internal/types"
 	"github.com/hashicorp/terraform-provider-aws/names"
 )
 
@@ -1768,93 +1761,4 @@ func flattenEKSVolumes(apiObjects []awstypes.EksVolume) []any {
 	}
 
 	return tfList
-}
-
-// @SDKListResource("aws_batch_job_definition")
-func jobDefinitionResourceAsListResource() inttypes.ListResourceForSDK {
-	l := jobDefinitionListResource{}
-	l.SetResourceSchema(resourceJobDefinition())
-	return &l
-}
-
-type jobDefinitionListResource struct {
-	framework.ResourceWithConfigure
-	framework.ListResourceWithSDKv2Resource
-	framework.ListResourceWithSDKv2Tags
-}
-
-type jobDefinitionListResourceModel struct {
-	framework.WithRegionModel
-}
-
-func (l *jobDefinitionListResource) ListResourceConfigSchema(ctx context.Context, request list.ListResourceSchemaRequest, response *list.ListResourceSchemaResponse) {
-	response.Schema = listschema.Schema{
-		Attributes: map[string]listschema.Attribute{},
-		Blocks:     map[string]listschema.Block{},
-	}
-}
-
-func (l *jobDefinitionListResource) List(ctx context.Context, request list.ListRequest, stream *list.ListResultsStream) {
-	awsClient := l.Meta()
-	conn := awsClient.BatchClient(ctx)
-
-	var query jobDefinitionListResourceModel
-	if request.Config.Raw.IsKnown() && !request.Config.Raw.IsNull() {
-		if diags := request.Config.Get(ctx, &query); diags.HasError() {
-			stream.Results = list.ListResultsStreamDiagnostics(diags)
-			return
-		}
-	}
-
-	var input batch.DescribeJobDefinitionsInput
-
-	tflog.Info(ctx, "Listing Batch job definitions")
-
-	stream.Results = func(yield func(list.ListResult) bool) {
-		pages := batch.NewDescribeJobDefinitionsPaginator(conn, &input)
-		for pages.HasMorePages() {
-			page, err := pages.NextPage(ctx)
-			if err != nil {
-				result := fwdiag.NewListResultErrorDiagnostic(err)
-				yield(result)
-				return
-			}
-
-			for _, jobDef := range page.JobDefinitions {
-				arn := aws.ToString(jobDef.JobDefinitionArn)
-				ctx := tflog.SetField(ctx, logging.ResourceAttributeKey(names.AttrID), arn)
-
-				result := request.NewListResult(ctx)
-				rd := l.ResourceData()
-				rd.SetId(arn)
-
-				tflog.Info(ctx, "Reading Batch job definition")
-				diags := resourceJobDefinitionRead(ctx, rd, awsClient)
-				if diags.HasError() {
-					result = fwdiag.NewListResultErrorDiagnostic(fmt.Errorf("reading Batch job definition %s", arn))
-					yield(result)
-					return
-				}
-
-				err = l.SetTags(ctx, awsClient, rd)
-				if err != nil {
-					result = fwdiag.NewListResultErrorDiagnostic(err)
-					yield(result)
-					return
-				}
-
-				result.DisplayName = aws.ToString(jobDef.JobDefinitionName)
-
-				l.SetResult(ctx, awsClient, request.IncludeResource, &result, rd)
-				if result.Diagnostics.HasError() {
-					yield(result)
-					return
-				}
-
-				if !yield(result) {
-					return
-				}
-			}
-		}
-	}
 }

--- a/internal/service/batch/job_definition_list.go
+++ b/internal/service/batch/job_definition_list.go
@@ -1,0 +1,113 @@
+// Copyright IBM Corp. 2014, 2025
+// SPDX-License-Identifier: MPL-2.0
+
+package batch
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/batch"
+	"github.com/hashicorp/terraform-plugin-framework/list"
+	listschema "github.com/hashicorp/terraform-plugin-framework/list/schema"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-provider-aws/internal/errs/fwdiag"
+	"github.com/hashicorp/terraform-provider-aws/internal/framework"
+	"github.com/hashicorp/terraform-provider-aws/internal/logging"
+	inttypes "github.com/hashicorp/terraform-provider-aws/internal/types"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
+
+// @SDKListResource("aws_batch_job_definition")
+func jobDefinitionResourceAsListResource() inttypes.ListResourceForSDK {
+	l := jobDefinitionListResource{}
+	l.SetResourceSchema(resourceJobDefinition())
+	return &l
+}
+
+type jobDefinitionListResource struct {
+	framework.ResourceWithConfigure
+	framework.ListResourceWithSDKv2Resource
+	framework.ListResourceWithSDKv2Tags
+}
+
+type jobDefinitionListResourceModel struct {
+	framework.WithRegionModel
+}
+
+func (l *jobDefinitionListResource) ListResourceConfigSchema(ctx context.Context, request list.ListResourceSchemaRequest, response *list.ListResourceSchemaResponse) {
+	response.Schema = listschema.Schema{
+		Attributes: map[string]listschema.Attribute{},
+		Blocks:     map[string]listschema.Block{},
+	}
+}
+
+func (l *jobDefinitionListResource) List(ctx context.Context, request list.ListRequest, stream *list.ListResultsStream) {
+	awsClient := l.Meta()
+	conn := awsClient.BatchClient(ctx)
+
+	var query jobDefinitionListResourceModel
+	if request.Config.Raw.IsKnown() && !request.Config.Raw.IsNull() {
+		if diags := request.Config.Get(ctx, &query); diags.HasError() {
+			stream.Results = list.ListResultsStreamDiagnostics(diags)
+			return
+		}
+	}
+
+	var input batch.DescribeJobDefinitionsInput
+
+	tflog.Info(ctx, "Listing Batch job definitions")
+
+	stream.Results = func(yield func(list.ListResult) bool) {
+		pages := batch.NewDescribeJobDefinitionsPaginator(conn, &input)
+		for pages.HasMorePages() {
+			page, err := pages.NextPage(ctx)
+			if err != nil {
+				result := fwdiag.NewListResultErrorDiagnostic(err)
+				yield(result)
+				return
+			}
+
+			for _, jobDef := range page.JobDefinitions {
+				arn := aws.ToString(jobDef.JobDefinitionArn)
+				ctx := tflog.SetField(ctx, logging.ResourceAttributeKey(names.AttrID), arn)
+
+				result := request.NewListResult(ctx)
+				rd := l.ResourceData()
+				rd.SetId(arn)
+
+				tflog.Info(ctx, "Reading Batch job definition")
+				diags := resourceJobDefinitionRead(ctx, rd, awsClient)
+				if diags.HasError() {
+					result = fwdiag.NewListResultErrorDiagnostic(fmt.Errorf("reading Batch job definition %s", arn))
+					yield(result)
+					return
+				}
+				if rd.Id() == "" {
+					// Resource is logically deleted
+					continue
+				}
+
+				err = l.SetTags(ctx, awsClient, rd)
+				if err != nil {
+					result = fwdiag.NewListResultErrorDiagnostic(err)
+					yield(result)
+					return
+				}
+
+				result.DisplayName = aws.ToString(jobDef.JobDefinitionName)
+
+				l.SetResult(ctx, awsClient, request.IncludeResource, &result, rd)
+				if result.Diagnostics.HasError() {
+					yield(result)
+					return
+				}
+
+				if !yield(result) {
+					return
+				}
+			}
+		}
+	}
+}

--- a/internal/service/batch/job_definition_list_test.go
+++ b/internal/service/batch/job_definition_list_test.go
@@ -16,6 +16,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-testing/tfversion"
 	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
 	tfknownvalue "github.com/hashicorp/terraform-provider-aws/internal/acctest/knownvalue"
+	tfstatecheck "github.com/hashicorp/terraform-provider-aws/internal/acctest/statecheck"
 	"github.com/hashicorp/terraform-provider-aws/names"
 )
 
@@ -25,6 +26,9 @@ func TestAccBatchJobDefinition_List_basic(t *testing.T) {
 	resourceName1 := "aws_batch_job_definition.test[0]"
 	resourceName2 := "aws_batch_job_definition.test[1]"
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	arn1 := tfstatecheck.StateValue()
+	arn2 := tfstatecheck.StateValue()
 
 	acctest.ParallelTest(ctx, t, resource.TestCase{
 		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
@@ -42,7 +46,10 @@ func TestAccBatchJobDefinition_List_basic(t *testing.T) {
 					acctest.CtRName: config.StringVariable(rName),
 				},
 				ConfigStateChecks: []statecheck.StateCheck{
+					arn1.GetStateValue(resourceName1, tfjsonpath.New(names.AttrARN)),
 					statecheck.ExpectKnownValue(resourceName1, tfjsonpath.New(names.AttrARN), tfknownvalue.RegionalARNExact("batch", "job-definition/"+rName+"-0:1")),
+
+					arn2.GetStateValue(resourceName2, tfjsonpath.New(names.AttrARN)),
 					statecheck.ExpectKnownValue(resourceName2, tfjsonpath.New(names.AttrARN), tfknownvalue.RegionalARNExact("batch", "job-definition/"+rName+"-1:1")),
 				},
 			},
@@ -57,11 +64,11 @@ func TestAccBatchJobDefinition_List_basic(t *testing.T) {
 				},
 				QueryResultChecks: []querycheck.QueryResultCheck{
 					querycheck.ExpectIdentity("aws_batch_job_definition.test", map[string]knownvalue.Check{
-						names.AttrARN: knownvalue.NotNull(),
+						names.AttrARN: arn1.Value(),
 					}),
 
 					querycheck.ExpectIdentity("aws_batch_job_definition.test", map[string]knownvalue.Check{
-						names.AttrARN: knownvalue.NotNull(),
+						names.AttrARN: arn2.Value(),
 					}),
 				},
 			},

